subseas_flex_fcst/maproom.py

```diff
@@ -0,0 +1,516 @@
+
+# App
+
+CONFIG = pyaconf.load(os.environ["CONFIG"])
+
+SERVER = flask.Flask(__name__)
+APP = dash.Dash(
+    __name__,
+    server=SERVER,
+    external_stylesheets=[
+        dbc.themes.BOOTSTRAP,
+        "https://use.fontawesome.com/releases/v5.12.1/css/all.css",
+    ],
+    url_base_pathname=f"{PFX}/",
+    meta_tags=[
+        {"name": "description", "content": "Seasonal Forecast"},
+        {"name": "viewport", "content": "width=device-width, initial-scale=1.0"},
+    ],
+)
+APP.title = "Sub-Seasonal Forecast"
+
+APP.layout = layout.app_layout()
+
+
+@APP.callback(
+    Output("navbar-collapse", "is_open"),
+    Input("navbar-toggler", "n_clicks"),
+    State("navbar-collapse", "is_open"),
+)
+def toggle_navbar_collapse(n, is_open):
+    if n:
+        return not is_open
+    return is_open
+
+
+@APP.callback(
+    Output("map", "center"),
+    Output("latInput", "min"),
+    Output("latInput", "max"),
+    Output("lngInput", "min"),
+    Output("lngInput", "max"),
+    Output("latLab", "children"),
+    Output("lonLab", "children"),
+    Input("submitLatLng","n_clicks"),
+)
+def initialize(toto):
+    fcst_mu = cptio.open_cptdataset(Path(DATA_PATH, Path(CONFIG["forecast_mu_file"])))
+    center_of_the_map = [((fcst_mu.Y[0]+fcst_mu.Y[-1])/2).values, ((fcst_mu.X[0]+fcst_mu.X[-1])/2).values]
+    lat_res = np.around((fcst_mu.Y[0]-fcst_mu.Y[1]).values, decimals=10)
+    lat_min = str(np.around((fcst_mu.Y[-1]-lat_res/2).values, decimals=10))
+    lat_max = str(np.around((fcst_mu.Y[0]+lat_res/2).values, decimals=10))
+    lon_res = np.around((fcst_mu.X[1]-fcst_mu.X[0]).values, decimals=10)
+    lon_min = str(np.around((fcst_mu.X[0]-lon_res/2).values, decimals=10))
+    lon_max = str(np.around((fcst_mu.X[-1]+lon_res/2).values, decimals=10))
+    lat_label = lat_min+" to "+lat_max+" by "+str(lat_res)+"˚"
+    lon_label = lon_min+" to "+lon_max+" by "+str(lon_res)+"˚"
+    return center_of_the_map, lat_min, lat_max, lon_min, lon_max, lat_label, lon_label
+
+
+def get_coords(click_lat_lng):
+    if click_lat_lng is not None:
+        return click_lat_lng
+    else:
+        fcst_mu = cptio.open_cptdataset(Path(DATA_PATH, Path(CONFIG["forecast_mu_file"])))
+        return [(fcst_mu.Y[0].values+fcst_mu.Y[-1].values)/2, (fcst_mu.X[0].values+fcst_mu.X[-1].values)/2]
+
+
+def round_latLng(coord):
+    value = float(coord)
+    value = round(value, 1)
+    return value
+
+
+@APP.callback(Output("map", "click_lat_lng"), Input("submitLatLng","n_clicks"), State("latInput", "value"), State("lngInput", "value"))
+def inputCoords(n_clicks,latitude,longitude):
+    if latitude is None:
+        return None
+    else:
+        lat_lng = [latitude, longitude]
+        return lat_lng
+
+
+@APP.callback(Output("layers_group", "children"), Input("map", "click_lat_lng"))
+def map_click(click_lat_lng):
+    lat_lng = get_coords(click_lat_lng)
+
+    return dlf.Marker(
+        position=lat_lng, children=dlf.Tooltip("({:.3f}, {:.3f})".format(*lat_lng))
```

Comment:
So here... I wonder if it might be interesting to have a finer precision than the data resolution so that the user potentially sees what data points they get vs what point they clicked. It's not clear thought what resolution/precision to apply for the tooltip. Must be higher than the data. Maybe the background map resolution? do we know it?

Reply:
The background map is generated from vector shapes, so it doesn't have an underlying pixel resolution in any straightforward sense.

Reply:
Oops... that's where I meant to post my comment. So decimal precision equal to variable everywhere except for the tooltip of the map where we can add one level of precision?

Reply:
Snapping to the grid regardless of the input method (click or type) seems clearest to me.

Since the lat and lon input boxes are set up as `State` rather than `Input`, it shouldn't be a problem to overwrite the typed values with the nearest grid point when the submit button is clicked.